src/components/popover/popover.ts

```diff
@@ -1,112 +1,303 @@
+        el.style[property] = style[property];
+      }
+    });
+  }
+
+  private resetPopoverElStyle() {
+    const el = this.popoverRef.el;
+    if (!el) return;
+    el.style["max-height"] = el.style["max-width"] = "";
+    el.style["bottom"] = el.style["top"] = el.style["right"] = el.style["left"] = "";
+  }
+}
+
+Popover.props = {
+  anchorRect: Object,
+  containerRect: { type: Object, optional: true },
+  positioning: { type: String, optional: true },
+  maxWidth: { type: Number, optional: true },
+  maxHeight: { type: Number, optional: true },
+  verticalOffset: { type: Number, optional: true },
+  onMouseWheel: { type: Function, optional: true },
+  slots: Object,
+};
+
+abstract class PopoverHelper {
+  constructor(
+    protected anchorRect: Rect,
+    protected containerRect: Rect,
+    private propsMaxSize: Partial<DOMDimension>
+  ) {}
+
+  protected abstract get availableHeightUp(): number;
+  protected abstract get availableHeightDown(): number;
+  protected abstract get availableWidthRight(): number;
+  protected abstract get availableWidthLeft(): number;
+
+  /** Check if there is enough space for the popover to be rendered at the bottom of the anchorRect */
+  protected shouldRenderAtBottom(elementHeight: number): boolean {
+    if (elementHeight <= this.availableHeightDown) return true;
+
+    if (
+      elementHeight > this.availableHeightUp &&
+      this.availableHeightDown >= this.availableHeightUp
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  /** Check if there is enough space for the popover to be rendered at the right of the anchorRect */
+  protected shouldRenderAtRight(elementWidth: number): boolean {
+    if (elementWidth <= this.availableWidthRight) return true;
+
+    if (
+      elementWidth > this.availableWidthLeft &&
+      this.availableWidthRight >= this.availableWidthLeft
+    ) {
+      return true;
+    }
+
+    return false;
+  }
+
+  protected getMaxHeight(elementHeight: number) {
+    const shouldRenderAtBottom = this.shouldRenderAtBottom(elementHeight);
+    const computedMaxHeight = shouldRenderAtBottom
+      ? this.availableHeightDown
+      : this.availableHeightUp;
+
+    return this.propsMaxSize.height
+      ? Math.min(computedMaxHeight, this.propsMaxSize.height)
+      : computedMaxHeight;
+  }
+
+  protected getMaxWidth(elementWidth: number) {
+    const shouldRenderAtRight = this.shouldRenderAtRight(elementWidth);
+    const computedMaxWidth = shouldRenderAtRight
+      ? this.availableWidthRight
+      : this.availableWidthLeft;
+
+    return this.propsMaxSize.width
+      ? Math.min(computedMaxWidth, this.propsMaxSize.width)
+      : computedMaxWidth;
+  }
+
+  protected abstract getTopCoordinate(
+    actualElementHeight: number,
+    shouldRenderAtBottom: boolean
+  ): number;
+
+  protected abstract getLeftCoordinate(
+    actualElementWidth: number,
+    shouldRenderAtRight: boolean
+  ): number;
+
+  getCSSStyle(
```

Comment:
maybe just `getCSS` ? or `getElementCSS` ?

Also, we should harmonize our naming. Either `...Css` or `...CSS`. We currently have both.

Reply:
I think that `CSS` makes a bit more sense, but `Css` is prettier with CamelCase so I'll use that I think :)